File: 11. NQueenBacktracking.py
"""
Task: Implementation and Time Analysis of N-Queen Problem using Backtracking
"""
def is_safe(board, row, col, n):
    for i in range(col):
        if board[row][i]:
            return False
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j]:
            return False
    for i, j in zip(range(row, n), range(col, -1, -1)):
        if board[i][j]:
            return False
    return True
def solve_nq_util(board, col, n):
    if col == n:
        return True
    for i in range(n):
        if is_safe(board, i, col, n):
            board[i][col] = 1
            if solve_nq_util(board, col+1, n):
                return True
            board[i][col] = 0
    return False
def print_board(board, n):
    for row in board:
        print(' '.join(str(x) for x in row))
if __name__ == "__main__":
    import time
    n = 4
    board = [[0]*n for _ in range(n)]
    start = time.time()
    if solve_nq_util(board, 0, n):
        print_board(board, n)
    else:
        print("No solution")
    end = time.time()
    print(f"Time: {end-start:.6f}s")
"""
Time Complexity: O(N!)
Space Complexity: O(N^2)
"""

File: 11. NQueenBacktracking.cpp
/*
Task: Implementation and Time Analysis of N-Queen Problem using Backtracking
*/
#include <iostream>
#include <vector>
#include <ctime>
using namespace std;

bool isSafe(vector<vector<int>>& board, int row, int col, int n)
{
    for (int i = 0; i < col; ++i)
        if (board[row][i])
            return false;
    for (int i = row, j = col; i >= 0 && j >= 0; --i, --j)
        if (board[i][j])
            return false;
    for (int i = row, j = col; i < n && j >= 0; ++i, --j)
        if (board[i][j])
            return false;
    return true;
}
bool solveNQUtil(vector<vector<int>>& board, int col, int n)
{
    if (col == n)
        return true;
    for (int i = 0; i < n; ++i)
    {
        if (isSafe(board, i, col, n))
        {
            board[i][col] = 1;
            if (solveNQUtil(board, col + 1, n))
                return true;
            board[i][col] = 0;
        }
    }
    return false;
}
void printBoard(const vector<vector<int>>& board, int n)
{
    for (int i = 0; i < n; ++i)
    {
        for (int j = 0; j < n; ++j) cout << board[i][j] << " ";
        cout << "\n";
    }
}
int main()
{
    int n = 4;
    vector<vector<int>> board(n, vector<int>(n, 0));
    clock_t start = clock();
    if (solveNQUtil(board, 0, n))
        printBoard(board, n);
    else
        cout << "No solution\n";
    clock_t end = clock();
    cout << "Time: " << (double)(end - start) / CLOCKS_PER_SEC << "s\n";
    return 0;
}
/*
Time Complexity: O(N!)
Space Complexity: O(N^2)
*/
